### PR TITLE
Improved request errors

### PR DESCRIPTION
**Pull request by biotom**
## Description
Errors returned in the file "Requests.go" were wrapped with useful error messages, and defer statements were added to http requests. Some extra error handling was added to "Requests_test.go".
## Motivation and Context
Requests.go returns unwrapped errors (often from common 3rd party libraries), so adding wrapping to those errors might provide more context and make troubleshooting easier. Additionally, 
deferred calls to Close() were added after http requests to prevent resource leakage. Require.Noerror statements were added to requests_test.go to handle unhandled errors.

## How Has This Been Tested?
Existing unit tests passed, and the code shouldn't cause changes to public behaviour. Error messages are wrapped to provide context, but the testing suite does not rely on literal matching of error messages. 

## Checklist:

- [x] My change requires a change to the documentation or CHANGELOG.
- [x] I have updated the documentation/CHANGELOG accordingly.
- [x] I have created a feature (non-master) branch for my PR.

### CODE REVIEW

File: pkg/requests/requests.go [hunk 1]
```diff
@@ -3,11 +3,11 @@ package requests
...
 	"encoding/json"
 	"fmt"
-	"io/ioutil"
-	"net/http"
-
 	"github.com/bitly/go-simplejson"
+	"github.com/pkg/errors"
 	"github.com/pusher/oauth2_proxy/pkg/logger"
+	"io/ioutil"
+	"net/http"
```
Comment (syscll) [inline pkg/requests/requests.go]: These imports are now out of order.

File: pkg/requests/requests_test.go [hunk 2]
```diff
@@ -1,6 +1,7 @@
 package requests
 
 import (
+	"github.com/stretchr/testify/require"
```
Comment (syscll) [inline pkg/requests/requests_test.go]: This import is out of order.

File: pkg/requests/requests.go [hunk 3]
```diff
@@ -41,10 +47,13 @@ func RequestJSON(req *http.Request, v interface{}) error {
 		return err
 	}
 	body, err := ioutil.ReadAll(resp.Body)
-	resp.Body.Close()
+	if body != nil {
+		resp.Body.Close()
```
Comment (syscll) [inline pkg/requests/requests.go]:
```suggestion
		defer resp.Body.Close()
```

**Comment by syscll** [comment]
While I agree with making these errors more verbose is a good idea, what advantage do we get over just doing:
```
fmt.Errorf("something bad happened: %v", err)
```

Am I right in thinking the callers of these functions would then need to parse the "new" errors to check for [`Cause()`](https://github.com/pkg/errors/blob/master/errors.go#L160)?

**Comment by syscll** [comment]
We could also be super fancy and start to implement some of the new go1.13 error functions https://blog.golang.org/go1.13-errors

File: pkg/requests/requests.go (hunk 1 above)
Comment (biotom) [reply pkg/requests/requests.go]: What import sorting scheme do you use? Goland complains if I swap these around, so I think they're ok according to the gofmt standard.

Comment (biotom) [reply pkg/requests/requests.go]: Here's how they look now: 
import (
	"encoding/json"
	"fmt"
	"github.com/bitly/go-simplejson"
	"github.com/pkg/errors"
	"github.com/pusher/oauth2_proxy/pkg/logger"
	"io/ioutil"
	"net/http"
)import (
	"encoding/json"
	"fmt"
	"github.com/bitly/go-simplejson"
	"github.com/pkg/errors"
	"github.com/pusher/oauth2_proxy/pkg/logger"
	"io/ioutil"
	"net/http"
)

Comment (syscll) [reply pkg/requests/requests.go]: We use [`goimports`](https://godoc.org/golang.org/x/tools/cmd/goimports) to format and auto include imports in this project.

According to the spec, the import block should be as follows:
```
import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"net/http"

	"github.com/bitly/go-simplejson"
	"github.com/pkg/errors"
	"github.com/pusher/oauth2_proxy/pkg/logger"
)
```

Third party imports should be after std lib imports and I think this is why CI is currently failing.

**Comment by biotom** [comment]
> While I agree with making these errors more verbose is a good idea, what advantage do we get over just doing:
> 
> ```
> fmt.Errorf("something bad happened: %v", err)
> ```
> 
> Am I right in thinking the callers of these functions would then need to parse the "new" errors to check for [`Cause()`](https://github.com/pkg/errors/blob/master/errors.go#L160)?


For me, the big advantage of Wrap is that you're automatically putting together the same sort of useful error string that you'd get with Errorf, but you can easily string a bunch of them together, making it easier to work out what went wrong for debugging. You also get a stack trace from the point at which Wrap was called. This extra context is useful when you're dealing with errors from 3rd party libraries, which might return a generic and unhelpful error message. You'd want to call Cause() further up the error chain if you want to do something based on what the original error was- but in the spirit of using opaque errors, and handling them based on behaviour, I haven't used it very often. 

I know error handling in Go is a bit of a can of worms, so I could be wrong about all this...

**Comment by biotom** [comment]
Apologies- I'm not really sure what I've broken here. Travis is failing with "go: golang.org/x/oauth2@v0.0.0-20190604053449-0f29369cfe45: unrecognized import path "golang.org/x/oauth2" (https fetch: Get https://golang.org/x/oauth2?go-get=1: dial tcp 74.125.69.141:443: i/o timeout)", but go mod tidy adds it to go.sum again when it is removed.

**Comment by syscll** [comment]
>so I could be wrong about all this...

You are **definitely not** wrong! I think this PR and your feedback is extremely useful.

My only "concern" is that http://github.com/pkg/errors was just a stop gap until we had better error handling in the Go std lib. For example, their [roadmap](https://github.com/pkg/errors#roadmap) says they won't cut anymore releases after v1.0 

>1.0. Final release.

As of Go 1.13, there is a new `Unwrap()` func in the std lib `errors` package. Here is a [playground](https://play.golang.org/p/uAVju1-mMrH) with an example.

My thoughts are, we could start implementing this new functionality instead of importing a new 3rd party package. WDYT @biotom @JoelSpeed?

**Comment by biotom** [comment]
> My thoughts are, we could start implementing this new functionality instead of importing a new 3rd party package. WDYT @biotom @JoelSpeed?

I think this makes a lot of sense! It looks to me like the new error handling is mostly the best wrapping bits from the errors library at any rate- I'd be happy to look into it, if people feel like that's a good idea.

**Comment by syscll** [comment]
I'd be happy to implement this new functionality with you if you want to work on it together? I can create an issue which outlines our intentions and a list of changed we need to make.

**Comment by biotom** [comment]
> I'd be happy to implement this new functionality with you if you want to work on it together? I can create an issue which outlines our intentions and a list of changed we need to make.

Sounds great!

**Comment by JoelSpeed** [comment]
> As of Go 1.13, there is a new Unwrap() func in the std lib errors package. Here is a playground with an example.

Out of interest, if this is in 1.13, can we not just use 1.13? 🤔

**Comment by syscll** [comment]
>Out of interest, if this is in 1.13, can we not just use 1.13? :thinking: 

@JoelSpeed I'm not sure what you mean by this?

There is now a new `%w` verb that can be passed to `fmt.Errorf` which automatically wraps errors, so we would need to implement this.

https://golang.org/src/fmt/errors.go

**Comment by biotom** [comment]
> > Out of interest, if this is in 1.13, can we not just use 1.13? 🤔
> 
> @JoelSpeed I'm not sure what you mean by this?
> 
> There is now a new `%w` verb that can be passed to `fmt.Errorf` which automatically wraps errors, so we would need to implement this.
> 
> https://golang.org/src/fmt/errors.go

Maybe a better one would be Wrapf(), since it adds some more context while also letting you use formatting in the error message?

File: pkg/requests/requests.go [hunk 4]
```diff
@@ -7,6 +7,7 @@ import (
 	"net/http"
 
 	"github.com/bitly/go-simplejson"
+
```
Comment (syscll) [inline pkg/requests/requests.go]: Don't need this newline.

File: pkg/requests/requests.go [hunk 5]
```diff
@@ -56,7 +66,7 @@ func RequestJSON(req *http.Request, v interface{}) error {
 func RequestUnparsedResponse(url string, header http.Header) (resp *http.Response, err error) {
 	req, err := http.NewRequest("GET", url, nil)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("error performing get request")
```
Comment (syscll) [inline pkg/requests/requests.go]:
```suggestion
		return nil, fmt.Errorf("error performing get request: %w", err)
```

File: pkg/requests/requests_test.go [hunk 6]
```diff
@@ -7,21 +7,24 @@ import (
 	"strings"
 	"testing"
 
+	"github.com/stretchr/testify/require"
```
Comment (syscll) [inline pkg/requests/requests_test.go]: This import needs to go after L14

File: pkg/requests/requests_test.go [hunk 7]
```diff
@@ -7,21 +7,24 @@ import (
...
+
 	"github.com/bitly/go-simplejson"
 
 	"github.com/stretchr/testify/assert"
 )
 
-func testBackend(responseCode int, payload string) *httptest.Server {
+func testBackend(responseCode int, payload string, t *testing.T) *httptest.Server {
```
Comment (syscll) [inline pkg/requests/requests_test.go]: It's convention to pass `t` as the first param to test functions.

**Comment by biotom** [comment]
> Thank you for you patience and input on this PR! Congrats on your first contribution 🎉

Thanks! And thanks for all the support with my first open source contribution! I'd be happy to help check out error handling in the rest of the code, if you'd still be interested.

**Comment by syscll** [comment]
@biotom I plan on creating an issue for a general refactor of errors as mentioned above. I'll add you as an assignee and you can help out whenever suits you best :smile: 

Also, we have an #oauth2_proxy channel in [Gophers Slack](https://gophers.slack.com) if you'd like to get involved in other conversations.

**Pull request merged**